source/code/plugins/out_oms.rb

```diff
@@ -121,17 +114,34 @@ def write(chunk)
         end
       }
 
-      datatypes.each do |key, record|
-        handle_record(key, record)
+      ret = []
+      [datatypes, unmergable_records].each do |list_records|
+        list_records.each { |key, records|
+          ret << {'source': key, 'event': handle_record(key, records)}
+        }
       end
 
-      @log.trace "Handling #{unmergable_records.size} unmergeable records"
-      unmergable_records.each { |key, record|
-        handle_record(key, record)
-      }
+      ret
     end
 
-  private
+    # This method is called every flush interval. Send the buffer chunk to OMS.
+    # 'chunk' is a buffer chunk that includes multiple formatted
+    # NOTE! This method is called by (out_oms) plugin thread not Fluentd's main thread. So IO wait doesn't affect other plugins.
+    def write(chunk)
+      # Quick exit if we are missing something
+      if !OMS::Configuration.load_configuration(omsadmin_conf_path, cert_path, key_path)
+        raise OMS::RetryRequestException, 'Missing configuration. Make sure to onboard.'
+      end
+
+      if run_in_background
+        OMS::BackgroundJobs.instance.run_job_and_wait { self_write(chunk) }
```

Comment:
are we sure there won't be any synchronization issues? for example, thread safety of "chunk"?

Reply:
We upload that chunk in another process, so there is no issue with synchronization.
We are going to enable this feature incrementally.